File: RadixCyclicNN/radixnet/blame.py
"""Where the negative network's data comes from: the tutor's verdicts, turned into blame.

Nothing in :class:`~radixnet.negative.NegativeNet` is invented.  Every failure
it knows was handed to it by something outside the network that looked at an
output and said it was wrong, and why:

* the **English tutor** (:mod:`radixnet.tutor`): the LLM sets an exercise, the
  network completes it and the LLM marks the sentence - a mark out of 10, the
  single worst mistake named out of :data:`radixnet.tutor.ERROR_TYPES`
  (``agreement``, ``tense``, ``article``, ``plural``, ...), one sentence of
  teaching and the *correction*, the same sentence written out in correct
  English.  The mistake is the reason, the mark is the severity, and the diff
  against the correction says which characters were wrong
  (:func:`faults_from_lessons`, :meth:`radixnet.negative.NegativeNet.correct`);
* the **adversarial reviewer** (:mod:`radixnet.ollama`): an Ollama model rates
  the network's own texts 0-10, passes or fails each one and writes a
  one-sentence critique;
* the **code-generation teacher and judge** (:mod:`radixnet.codegen`): the
  sandbox says a program crashed, timed out or printed the wrong thing, the
  style checker names its issues and the LLM judge says whether the task was
  accomplished at all;
* a **person** pressing thumbs down in the frontend, and the evolve loop's
  discriminator, which is the network's own critic.

This module turns those verdicts into **faults**.  A fault is a text, a
**reason** (a short tag - the tutor's own error type, or one out of
:data:`REASONS` / :data:`CODE_REASONS` picked from the tutor's words by
:func:`classify`), a **severity** (how badly it failed) and a **note** (the
tutor's sentence, kept verbatim for the journal).  :func:`teach` hands faults
to the negative network; the texts the tutor *passed* are cleared in the same
call, so a fragment that shows up in good and bad output alike stops carrying
the verdict on its own.
"""

from __future__ import annotations

from collections.abc import Iterable, Sequence
from typing import Any

__all__ = [
    "CODE_REASONS",
    "CODE_SEVERITY",
    "DEFAULT_REASON",
    "REASONS",
    "classify",
    "code_reason",
    "faults_from_attempts",
    "faults_from_lessons",
    "faults_from_reviews",
    "severity_from_rating",
    "teach",
    "teach_attempts",
    "teach_lessons",
    "teach_reviews",
]

DEFAULT_REASON = "other"

REASONS = (
    "empty",
    "gibberish",
    "repetition",
    "truncated",
    "grammar",
    "spelling",
    "contradiction",
    "false",
    "incoherent",
    "off-topic",
    DEFAULT_REASON,
)
"""Reason tags for reviewed *text* (the order is the order :func:`classify` tries them in)."""

CODE_REASONS = (
    "timeout",
    "crash",
    "wrong-output",
    "task-not-done",
    "style",
    "naming",
    DEFAULT_REASON,
)
"""Reason tags for reviewed *programs*."""

CODE_SEVERITY = {
    "timeout": 1.5,
    "crash": 1.5,
    "wrong-output": 1.25,
    "task-not-done": 1.0,
    "style": 0.5,
    "naming": 0.5,
    DEFAULT_REASON: 1.0,
}
"""How heavily each code failure is blamed (1 = one ordinary failure)."""

_PATTERNS: tuple[tuple[str, tuple[str, ...]], ...] = (
    ("empty", ("empty output", "no output", "produced nothing", "blank")),
    ("gibberish", ("gibberish", "nonsense word", "word salad", "garbled", "random character", "not words", "noise")),
    ("repetition", ("repeat", "repetit", "duplicat", "over and over", "loops")),
    ("truncated", ("truncat", "cut off", "cut short", "incomplete", "unfinished", "mid-sentence", "mid sentence")),
    ("grammar", ("grammar", "grammatic", "syntax", "word order", "punctuation", "agreement", "tense", "malformed sentence")),
    ("spelling", ("spelling", "misspell", "typo")),
    ("contradiction", ("contradict", "inconsisten", "conflicts with")),
    ("false", ("false", "factual", "inaccurate", "untrue", "wrong fact", "not true", "misleading")),
    ("incoherent", ("incoheren", "meaningless", "make no sense", "does not make sense", "doesn't make sense",
                    "nonsensical", "confusing", "unintelligible")),
    ("off-topic", ("off-topic", "off topic", "irrelevant", "unrelated", "does not answer", "ignores the prompt")),
)


def classify(critique: str | None, *, verdict: str | None = None, rating: float | None = None,
             default: str = DEFAULT_REASON) -> str:
    """The reason tag behind a tutor's critique (its own words decide; ``default`` when nothing matches).

    ``"unrated"`` comes back when the tutor failed a text without a critique
    the vocabulary recognises - it still failed, and the network records that
    nobody said why.
    """
    text = " ".join(str(critique or "").split()).lower()
    for reason, needles in _PATTERNS:
        if any(needle in text for needle in needles):
            return reason
    if (verdict or "").strip().lower() == "unrated":
        return "unrated"  # the tutor failed it without saying anything the vocabulary knows
    if rating is not None and rating <= 0 and not text:
        return "gibberish"
    return default


def severity_from_rating(rating: float | None, threshold: float = 6.0, floor: float = 0.25,
                         ceiling: float = 2.0) -> float:
    """How heavily a rated failure is blamed: ``ceiling`` at rating 0, ``floor`` at the pass threshold.

    An unrated failure (``rating`` ``None``) is blamed like one ordinary
    failure (1.0).
    """
    if rating is None:
        return 1.0
    limit = float(threshold) if threshold > 0 else 1.0
    share = max(0.0, min(1.0, (limit - float(rating)) / limit))
    return floor + (ceiling - floor) * share


def code_reason(attempt: Any) -> str:
    """The reason a code attempt was rejected, from the sandbox, the style report and the judge."""
    data = attempt.to_dict() if hasattr(attempt, "to_dict") else dict(attempt or {})
    run = data.get("run") or {}
    style = data.get("style") or {}
    verdict = data.get("verdict") or {}
    if run.get("timed_out"):
        return "timeout"
    if run.get("ok") is False:
        return "crash"
    if run.get("expected_ok") is False:
        return "wrong-output"
    if verdict.get("task") is False:
        return "task-not-done"
    if verdict.get("naming") is False or style.get("naming_ok") is False:
        return "naming"
    if verdict.get("pep8") is False or style.get("pep8_ok") is False:
        return "style"
    critique = verdict.get("critique") or ""
    issues = verdict.get("issues") or []
    return classify(critique or "; ".join(str(i) for i in issues[:3]))


def _fault(text: str, reason: str, severity: float, note: str, source: str) -> dict:
    return {"text": text, "reason": reason, "severity": float(severity), "note": note, "source": source}


def faults_from_reviews(reviews: Iterable[dict], threshold: float = 6.0, source: str = "review") -> tuple[list[dict], list[str]]:
    """``(faults, passed_texts)`` from :func:`radixnet.ollama.review_texts` entries.

    Everything the reviewer did not pass becomes a fault whose reason comes
    from its critique and whose severity comes from its rating; the texts it
    passed come back separately so they can clear blame.
    """
    faults: list[dict] = []
    passed: list[str] = []
    for entry in reviews or []:
        if not isinstance(entry, dict):
            continue
        text = str(entry.get("text") or "")
        if not text:
            continue
        verdict = str(entry.get("verdict") or "").strip().lower()
        rating = entry.get("rating")
        rating = float(rating) if isinstance(rating, (int, float)) else None
        if verdict == "pass":
            passed.append(text)
            continue
        critique = str(entry.get("critique") or "")
        faults.append(_fault(
            text,
            classify(critique, verdict=verdict, rating=rating),
            severity_from_rating(rating, threshold),
            critique,
            source,
        ))
    return faults, passed


def faults_from_attempts(attempts: Iterable[Any], source: str = "codegen") -> tuple[list[dict], list[str]]:
    """``(faults, correct_texts)`` from :class:`radixnet.codegen.Attempt` objects (or their dicts).

    A rejected program is blamed for what the sandbox, the style checker or
    the judge found (:func:`code_reason`), with the tutor's feedback as the
    note; the programs that were accepted come back to clear blame.
    """
    faults: list[dict] = []
    correct: list[str] = []
    for attempt in attempts or []:
        data = attempt.to_dict() if hasattr(attempt, "to_dict") else dict(attempt or {})
        text = getattr(attempt, "text", None) or data.get("text") or data.get("code") or ""
        text = str(text)
        if not text:
            continue
        if data.get("correct") or (data.get("verdict") or {}).get("correct"):
            correct.append(text)
            continue
        reason = code_reason(attempt)
        note = attempt.feedback() if hasattr(attempt, "feedback") else "; ".join(
            str(i) for i in ((data.get("verdict") or {}).get("issues") or [])[:4]
        )
        faults.append(_fault(text, reason, CODE_SEVERITY.get(reason, 1.0), note, source))
    return faults, correct


def faults_from_lessons(lessons: Iterable[Any], threshold: float = 6.0, source: str = "tutor") -> tuple[list[dict], list[str]]:
    """``(faults, passed_texts)`` from :class:`radixnet.tutor.Lesson` objects (or their dicts).

    The English tutor is the richest source of negatives there is: the mistake
    it named (``grade.error``, one of :data:`radixnet.tutor.ERROR_TYPES`) is
    the reason, its mark is the severity, its sentence of teaching is the note
    and its **correction** rides along in the fault, so :func:`teach` can blame
    only the characters the teacher actually changed
    (:meth:`radixnet.negative.NegativeNet.correct`).  The sentences that
    passed, the corrections themselves and the teacher's own model answers all
    come back as cleared text.
    """
    faults: list[dict] = []
    passed: list[str] = []
    for lesson in lessons or []:
        data = lesson.to_dict() if hasattr(lesson, "to_dict") else dict(lesson or {})
        sentence = " ".join(str(data.get("sentence") or "").split())
        grade = data.get("grade") or {}
        exercise = data.get("exercise") or {}
        correction = " ".join(str(grade.get("correction") or "").split())
        answer = " ".join(str(exercise.get("answer") or "").split())
        if sentence and grade.get("passed"):
            passed.append(sentence)
        elif sentence:
            error = str(grade.get("error") or "").strip().lower()
            comment = str(grade.get("comment") or "")
            reason = error if error and error != "none" else classify(comment, verdict=grade.get("graded_by"))
            score = grade.get("score")
            fault = _fault(
                sentence,
                reason,
                severity_from_rating(score if isinstance(score, (int, float)) else None, threshold),
                comment,
                source,
            )
            if correction and correction != sentence:
                fault["correction"] = correction
            faults.append(fault)
        for text in (correction, answer):
            if text and text not in passed:
                passed.append(text)
    return faults, passed


def teach(
    negative: Any,
    faults: Sequence[dict],
    passed: Sequence[str] = (),
    *,
    epochs: int = 1,
    clear_epochs: int = 1,
    progress: Any = None,
    stop_event: Any = None,
) -> dict:
    """Hand the tutor's faults to the negative network; ``passed`` texts clear blame afterwards.

    Returns ``{"blamed", "cleared", "unmatched", "edges", "reasons",
    "severity_mean", "records"}``.
    """
    records: list[dict] = []
    reasons: dict[str, int] = {}
    edges = 0
    severities: list[float] = []
    for fault in faults:
        if stop_event is not None and stop_event.is_set():
            break
        text = str(fault.get("text") or "")
        if not text:
            continue
        reason = str(fault.get("reason") or DEFAULT_REASON)
        severity = float(fault.get("severity", 1.0))
        correction = str(fault.get("correction") or "")
        source = str(fault.get("source") or "tutor")
        note = str(fault.get("note") or "")
        if correction and hasattr(negative, "correct"):
            # the tutor wrote the sentence out correctly: blame only the characters it changed
            outcome = negative.correct(
                text, correction, reason=reason, severity=severity, source=source, note=note,
            )
            edges += outcome["blamed"]
            reasons[reason] = reasons.get(reason, 0) + 1
            severities.append(severity)
            records.append(outcome)
            continue
        out = negative.blame(
            [text], reason=reason, severity=severity, source=source, note=note,
            epochs=epochs, progress=progress, stop_event=stop_event,
        )
        if out:
            edges += out[-1].get("edges_touched", 0)
        records.extend(out)
        reasons[reason] = reasons.get(reason, 0) + 1
        severities.append(severity)
    cleared = 0
    unmatched = 0
    texts = [str(t) for t in passed if str(t)]
    if texts and not (stop_event is not None and stop_event.is_set()):
        out = negative.clear(texts, epochs=clear_epochs, progress=progress, stop_event=stop_event)
        records.extend(out)
        if out:
            cleared = out[-1].get("matched", 0)
            unmatched = out[-1].get("unmatched", 0)
    return {
        "blamed": len(severities),
        "cleared": cleared,
        "unmatched": unmatched,
        "edges": edges,
        "reasons": reasons,
        "severity_mean": sum(severities) / len(severities) if severities else 0.0,
        "records": records,
    }


def _reviews_of(reviews: Any) -> tuple[list[dict], float | None]:
    """Accept the ``reviews`` list or the whole :func:`radixnet.ollama.adversarial_review` result."""
    if isinstance(reviews, dict):
        return list(reviews.get("reviews") or []), reviews.get("threshold")
    return list(reviews or []), None


def teach_reviews(negative: Any, reviews: Any, *, threshold: float = 6.0, clear_passes: bool = True,
                  source: str = "review", **options: Any) -> dict:
    """Feed an adversarial review straight into the negative network (see :func:`teach`)."""
    entries, own = _reviews_of(reviews)
    limit = float(own) if isinstance(own, (int, float)) else float(threshold)
    faults, passed = faults_from_reviews(entries, limit, source)
    report = teach(negative, faults, passed if clear_passes else (), **options)
    report.update(source=source, threshold=limit, faults=faults, passed=len(passed))
    return report


def teach_lessons(negative: Any, lessons: Iterable[Any], *, threshold: float = 6.0, clear_passes: bool = True,
                  source: str = "tutor", **options: Any) -> dict:
    """Feed a round of English lessons into the negative network (see :func:`faults_from_lessons`)."""
    faults, passed = faults_from_lessons(lessons, threshold, source)
    report = teach(negative, faults, passed if clear_passes else (), **options)
    report.update(source=source, threshold=float(threshold), faults=faults, passed=len(passed))
    return report


def teach_attempts(negative: Any, attempts: Iterable[Any], *, clear_passes: bool = True, source: str = "codegen",
                   **options: Any) -> dict:
    """Feed the code-generation teacher's rejected attempts into the negative network (see :func:`teach`)."""
    faults, correct = faults_from_attempts(attempts, source)
    report = teach(negative, faults, correct if clear_passes else (), **options)
    report.update(source=source, faults=faults, passed=len(correct))
    return report

File: RadixCyclicNN/rust/src/blame.rs
//! Where the negative network's data comes from: the verdicts of whatever
//! looked at an output and said it was wrong, turned into blame
//! (`radixnet/blame.py`, `go/radixnet/blame.go`).
//!
//! Nothing in the negative network is invented.  A [`Fault`] is a text, a
//! reason (the tutor's own error type, or one picked from its words by
//! [`classify`]), a severity (how badly it failed) and a note (the tutor's own
//! sentence, kept verbatim for the journal).  [`teach`] hands faults to the
//! negative network; the texts the tutor passed are cleared in the same call,
//! so a fragment that shows up in good and bad output alike stops carrying the
//! verdict on its own.
//!
//! This module is the core every tutor shares.  Each tutor turns its own
//! records into faults beside its own types: the review in `review.rs`, the
//! English lessons in `tutor.rs`, program attempts in `codegen.rs`, tool use
//! in `agent.rs`, recall in `recall.rs`.

use crate::json::Json;
use crate::model::{EpochRecord, Model};
use crate::negative::BlameOptions;

/// Recorded when nothing in the tutor's words is recognised.
pub const DEFAULT_REASON: &str = "other";

/// Reason tags for reviewed *text* (the order [`classify`] tries them in).
pub const REASONS: &[&str] = &[
    "empty",
    "gibberish",
    "repetition",
    "truncated",
    "grammar",
    "spelling",
    "contradiction",
    "false",
    "incoherent",
    "off-topic",
    DEFAULT_REASON,
];

/// Reason tags for reviewed *programs*.
pub const CODE_REASONS: &[&str] = &[
    "timeout",
    "crash",
    "wrong-output",
    "task-not-done",
    "style",
    "naming",
    DEFAULT_REASON,
];

/// How heavily each code failure is blamed (1 = one ordinary failure).
pub const CODE_SEVERITY: &[(&str, f64)] = &[
    ("timeout", 1.5),
    ("crash", 1.5),
    ("wrong-output", 1.25),
    ("task-not-done", 1.0),
    ("style", 0.5),
    ("naming", 0.5),
    (DEFAULT_REASON, 1.0),
];

/// Reason tags for an attempt at a task with *tools* (the order the agent's
/// reason is tried in).
pub const AGENT_REASONS: &[&str] = &["no-call", "bad-call", "tool-error", "no-answer", DEFAULT_REASON];

/// How heavily each tool-use failure is blamed.
pub const AGENT_SEVERITY: &[(&str, f64)] = &[
    ("no-call", 1.0),
    // one step of one attempt, not the whole attempt
    ("bad-call", 0.75),
    ("tool-error", 1.0),
    ("no-answer", 1.5),
    (DEFAULT_REASON, 1.0),
];

/// Reason tags for a waveform the network was asked to remember.
pub const SPEECH_REASONS: &[&str] = &[
    "unreadable",
    "truncated",
    "overrun",
    "garbled",
    "silence",
    "clipping",
    "mishearing",
    "distortion",
];

/// Reason tags for an image the network was asked to remember.
pub const IMAGE_REASONS: &[&str] = &[
    "unreadable",
    "truncated",
    "overrun",
    "garbled",
    "blank",
    "noise",
    "drift",
];

/// Payload length ratios outside which a recalled text is short or long
/// rather than merely wrong.
pub const RECALL_TRUNCATED: f64 = 0.9;
pub const RECALL_OVERRUN: f64 = 1.1;

/// Below this payload agreement a readable, right-length recall is still a
/// distortion (an image: a drift).
pub const RECALL_AGREEMENT: f64 = 0.9;

const PATTERNS: &[(&str, &[&str])] = &[
    ("empty", &["empty output", "no output", "produced nothing", "blank"]),
    (
        "gibberish",
        &[
            "gibberish",
            "nonsense word",
            "word salad",
            "garbled",
            "random character",
            "not words",
            "noise",
        ],
    ),
    (
        "repetition",
        &["repeat", "repetit", "duplicat", "over and over", "loops"],
    ),
    (
        "truncated",
        &[
            "truncat",
            "cut off",
            "cut short",
            "incomplete",
            "unfinished",
            "mid-sentence",
            "mid sentence",
        ],
    ),
    (
        "grammar",
        &[
            "grammar",
            "grammatic",
            "syntax",
            "word order",
            "punctuation",
            "agreement",
            "tense",
            "malformed sentence",
        ],
    ),
    ("spelling", &["spelling", "misspell", "typo"]),
    ("contradiction", &["contradict", "inconsisten", "conflicts with"]),
    (
        "false",
        &[
            "false",
            "factual",
            "inaccurate",
            "untrue",
            "wrong fact",
            "not true",
            "misleading",
        ],
    ),
    (
        "incoherent",
        &[
            "incoheren",
            "meaningless",
            "make no sense",
            "does not make sense",
            "doesn't make sense",
            "nonsensical",
            "confusing",
            "unintelligible",
        ],
    ),
    (
        "off-topic",
        &[
            "off-topic",
            "off topic",
            "irrelevant",
            "unrelated",
            "does not answer",
            "ignores the prompt",
        ],
    ),
];

/// The reason tag behind a tutor's critique: its own words decide, `default`
/// (else [`DEFAULT_REASON`]) covers what the vocabulary does not know.
///
/// `"unrated"` comes back when the tutor failed a text without a critique the
/// vocabulary recognises - it still failed, and the network records that
/// nobody said why.
pub fn classify(critique: &str, verdict: &str, rating: Option<f64>, default: &str) -> String {
    let text = critique.split_whitespace().collect::<Vec<_>>().join(" ").to_lowercase();
    for (reason, needles) in PATTERNS {
        if needles.iter().any(|n| text.contains(n)) {
            return reason.to_string();
        }
    }
    if verdict.trim().to_lowercase() == "unrated" {
        return "unrated".to_string();
    }
    if rating.is_some_and(|r| r <= 0.0) && text.is_empty() {
        return "gibberish".to_string();
    }
    if default.is_empty() {
        DEFAULT_REASON.to_string()
    } else {
        default.to_string()
    }
}

/// How heavily a rated failure is blamed: 2.0 at rating 0, 0.25 at the pass
/// threshold; an unrated failure (`None`) is blamed like one ordinary failure.
pub fn severity_from_rating(rating: Option<f64>, threshold: f64) -> f64 {
    const FLOOR: f64 = 0.25;
    const CEILING: f64 = 2.0;
    let Some(rating) = rating else { return 1.0 };
    let limit = if threshold > 0.0 { threshold } else { 1.0 };
    let share = ((limit - rating) / limit).clamp(0.0, 1.0);
    FLOOR + (CEILING - FLOOR) * share
}

/// How heavily a failure is blamed from how badly it failed (`gap` in
/// `[0, 1]`), on the scale [`severity_from_rating`] maps ratings onto.
pub fn severity_from_gap(gap: Option<f64>) -> f64 {
    const FLOOR: f64 = 0.25;
    const CEILING: f64 = 2.0;
    match gap {
        None => 1.0,
        Some(g) => FLOOR + (CEILING - FLOOR) * g.clamp(0.0, 1.0),
    }
}

/// The severity a table like [`CODE_SEVERITY`] gives a reason (1 when absent).
pub fn severity_of(table: &[(&str, f64)], reason: &str) -> f64 {
    table.iter().find(|(r, _)| *r == reason).map(|(_, s)| *s).unwrap_or(1.0)
}

/// One lesson for the negative network.
#[derive(Clone, Debug, PartialEq)]
pub struct Fault {
    pub text: String,
    pub reason: String,
    pub severity: f64,
    pub note: String,
    pub source: String,
    /// The sentence written out correctly, when the tutor gave one: then only
    /// what it changed is blamed.
    pub correction: String,
}

impl Fault {
    pub fn new(text: &str, reason: &str, severity: f64, note: &str, source: &str) -> Fault {
        Fault {
            text: text.to_string(),
            reason: reason.to_string(),
            severity,
            note: note.to_string(),
            source: source.to_string(),
            correction: String::new(),
        }
    }

    /// `{"text", "reason", "severity", "note", "source"}` (+ `correction`).
    pub fn to_json(&self) -> Json {
        let mut pairs = vec![
            ("text".to_string(), Json::str(self.text.clone())),
            ("reason".to_string(), Json::str(self.reason.clone())),
            ("severity".to_string(), Json::Num(self.severity)),
            ("note".to_string(), Json::str(self.note.clone())),
            ("source".to_string(), Json::str(self.source.clone())),
        ];
        if !self.correction.is_empty() {
            pairs.push(("correction".to_string(), Json::str(self.correction.clone())));
        }
        Json::Obj(pairs)
    }
}

/// What one round of blaming taught the negative network.
#[derive(Clone, Debug, Default)]
pub struct TeachReport {
    /// Faults blamed.
    pub blamed: usize,
    /// Passed texts that cleared something.
    pub cleared: usize,
    /// Passed texts the failure structure could not walk.
    pub unmatched: usize,
    /// Edges the blame touched.
    pub edges: usize,
    /// How many faults each reason had, in the order first met.
    pub reasons: Vec<(String, usize)>,
    pub severity_mean: f64,
    /// Every record: blame and clearing epochs, and correction outcomes.
    pub records: Vec<Json>,
    /// Filled in by the tutor that built the faults.
    pub source: String,
    pub threshold: Option<f64>,
    pub faults: Vec<Fault>,
    pub passed: usize,
}

impl TeachReport {
    /// `{"blamed", "cleared", "unmatched", "edges", "reasons",
    /// "severity_mean", "records"}` and, when a tutor filled them in,
    /// `source`, `threshold`, `faults` and `passed`.
    pub fn to_json(&self) -> Json {
        let mut pairs = vec![
            ("blamed".to_string(), Json::Int(self.blamed as i64)),
            ("cleared".to_string(), Json::Int(self.cleared as i64)),
            ("unmatched".to_string(), Json::Int(self.unmatched as i64)),
            ("edges".to_string(), Json::Int(self.edges as i64)),
            (
                "reasons".to_string(),
                Json::Obj(
                    self.reasons
                        .iter()
                        .map(|(r, n)| (r.clone(), Json::Int(*n as i64)))
                        .collect(),
                ),
            ),
            ("severity_mean".to_string(), Json::Num(self.severity_mean)),
            ("records".to_string(), Json::Arr(self.records.clone())),
        ];
        if !self.source.is_empty() {
            pairs.push(("source".to_string(), Json::str(self.source.clone())));
            pairs.push((
                "threshold".to_string(),
                self.threshold.map(Json::Num).unwrap_or(Json::Null),
            ));
            pairs.push((
                "faults".to_string(),
                Json::Arr(self.faults.iter().map(|f| f.to_json()).collect()),
            ));
            pairs.push(("passed".to_string(), Json::Int(self.passed as i64)));
        }
        Json::Obj(pairs)
    }
}

/// How one [`teach`] call runs.
pub struct TeachOptions<'a> {
    /// Blame passes per fault (0 = one).
    pub epochs: usize,
    /// Clearing passes over the passed texts (0 = one).
    pub clear_epochs: usize,
    /// Asked between faults and after every epoch; `true` stops the round.
    pub stop: Option<&'a (dyn Fn() -> bool + Sync)>,
}

impl Default for TeachOptions<'_> {
    fn default() -> Self {
        TeachOptions {
            epochs: 1,
            clear_epochs: 1,
            stop: None,
        }
    }
}

impl TeachOptions<'_> {
    fn stopped(&self) -> bool {
        self.stop.is_some_and(|stop| stop())
    }
}

/// Hands the tutor's faults to the negative network; the `passed` texts clear
/// blame afterwards.  A fault that carries a correction is taught from its
/// diff, so only the units the teacher changed are blamed.
pub fn teach(
    negative: &mut Model,
    faults: &[Fault],
    passed: &[String],
    o: &TeachOptions,
) -> Result<TeachReport, String> {
    if !negative.is_negative() {
        return Err(format!(
            "teaching needs the negative network (this is the {} model)",
            negative.kind()
        ));
    }
    let mut report = TeachReport::default();
    let mut severities: Vec<f64> = Vec::new();
    let count_reason =
        |reasons: &mut Vec<(String, usize)>, reason: &str| match reasons.iter_mut().find(|(r, _)| r == reason) {
            Some((_, n)) => *n += 1,
            None => reasons.push((reason.to_string(), 1)),
        };
    let record_json = |r: &EpochRecord| r.to_json();
    for fault in faults {
        if o.stopped() {
            break;
        }
        if fault.text.is_empty() {
            continue;
        }
        let reason = if fault.reason.is_empty() {
            DEFAULT_REASON.to_string()
        } else {
            fault.reason.clone()
        };
        let source = if fault.source.is_empty() {
            "tutor".to_string()
        } else {
            fault.source.clone()
        };
        let opts = BlameOptions {
            reason: reason.clone(),
            severity: fault.severity,
            source,
            note: fault.note.clone(),
            epochs: o.epochs,
            no_compress: false,
        };
        if !fault.correction.is_empty() {
            // the tutor wrote the sentence out correctly: blame only the units it changed
            let outcome = negative.blame_correction(&fault.text, &fault.correction, &opts, 1.0)?;
            report.edges += outcome.blamed;
            report.records.push(outcome.to_json());
        } else {
            let stop = o.stop;
            let records = negative.blame_with(std::slice::from_ref(&fault.text), &opts, &mut |_| {
                !stop.is_some_and(|s| s())
            })?;
            if let Some(last) = records.last() {
                report.edges += last.extra_value("edges_touched").and_then(|v| v.as_i64()).unwrap_or(0) as usize;
            }
            report.records.extend(records.iter().map(record_json));
        }
        count_reason(&mut report.reasons, &reason);
        severities.push(fault.severity);
    }
    let texts: Vec<String> = passed.iter().filter(|t| !t.is_empty()).cloned().collect();
    if !texts.is_empty() && !o.stopped() {
        let stop = o.stop;
        let records = negative.clear_with(&texts, 1.0, o.clear_epochs, &mut |_| !stop.is_some_and(|s| s()))?;
        if let Some(last) = records.last() {
            report.cleared = last.extra_value("matched").and_then(|v| v.as_i64()).unwrap_or(0) as usize;
            report.unmatched = last.extra_value("unmatched").and_then(|v| v.as_i64()).unwrap_or(0) as usize;
        }
        report.records.extend(records.iter().map(record_json));
    }
    report.blamed = severities.len();
    report.severity_mean = if severities.is_empty() {
        0.0
    } else {
        severities.iter().sum::<f64>() / severities.len() as f64
    };
    Ok(report)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::negative::NegativeOptions;

    #[test]
    fn the_critique_names_the_reason() {
        assert_eq!(classify("This is word salad.", "", None, ""), "gibberish");
        assert_eq!(classify("wrong TENSE here", "", None, ""), "grammar");
        assert_eq!(classify("", "unrated", None, ""), "unrated");
        assert_eq!(classify("", "", Some(0.0), ""), "gibberish");
        assert_eq!(classify("meh", "", Some(3.0), "thumbs-down"), "thumbs-down");
        assert_eq!(classify("meh", "", None, ""), DEFAULT_REASON);
    }

    #[test]
    fn a_worse_mark_is_a_heavier_blame() {
        assert_eq!(severity_from_rating(None, 6.0), 1.0);
        assert_eq!(severity_from_rating(Some(0.0), 6.0), 2.0);
        assert_eq!(severity_from_rating(Some(6.0), 6.0), 0.25);
        assert!(severity_from_rating(Some(2.0), 6.0) > severity_from_rating(Some(4.0), 6.0));
        assert_eq!(severity_from_gap(Some(1.0)), 2.0);
        assert_eq!(severity_of(CODE_SEVERITY, "crash"), 1.5);
        assert_eq!(severity_of(CODE_SEVERITY, "nope"), 1.0);
    }

    #[test]
    fn faults_blame_and_passes_clear() {
        let mut neg = Model::new_negative(0, &NegativeOptions::default()).unwrap();
        let mut corrected = Fault::new("he go to school", "grammar", 1.5, "go -> goes", "tutor");
        corrected.correction = "he goes to school".to_string();
        let faults = vec![
            Fault::new("the cat sat on the sky", "false", 1.0, "", "review"),
            corrected,
            Fault::new("", "empty", 1.0, "", "review"),
        ];
        let passed = vec!["the cat sat on the mat".to_string()];
        let report = teach(&mut neg, &faults, &passed, &TeachOptions::default()).unwrap();
        assert_eq!(report.blamed, 2, "an empty text is not a fault");
        assert!(report.edges > 0);
        assert_eq!(
            report.reasons,
            vec![("false".to_string(), 1), ("grammar".to_string(), 1)]
        );
        assert_eq!(report.cleared + report.unmatched, 1);
        assert!(
            report.cleared == 1,
            "the pass shares 'the cat sat on the' with a failure"
        );
        assert!((report.severity_mean - 1.25).abs() < 1e-12);
        let doc = report.to_json();
        assert_eq!(doc.at("blamed").as_i64(), Some(2));
        // the correction blamed the changed word, not the whole sentence
        let correction = report
            .records
            .iter()
            .find(|r| r.at("phase").as_str() == Some("correction"))
            .unwrap();
        assert!(correction.at("blamed").as_i64().unwrap() > 0);
        assert_eq!(correction.at("edits").as_i64(), Some(1));
    }

    #[test]
    fn a_positive_model_cannot_be_taught_failures() {
        let mut m = Model::new(0, crate::GraphOptions::default()).unwrap();
        assert!(teach(&mut m, &[], &[], &TeachOptions::default()).is_err());
    }
}
